If minifier_cmd fails, include output in fail msg

Enselic opened the pull request:
The output is likely to include helpful information.

In my case: Output before change:


`        Minibundle: Bundling js assets:
        Minibundle: js/main.js
  Liquid Exception: Bundling js assets failed with exit status 1, command: r.js -o baseUrl=js name=main out=stdout in index.html
`

Output after change, which contains information crucial to continue debugging the problem:

`        Minibundle: Bundling js assets:
        Minibundle: js/main.js
  Liquid Exception: Bundling js assets failed with exit status 1, command: r.js -o baseUrl=js name=main out=stdout output: Tracing dependencies for: main Uglifying file: main.build.js Error: TypeError: process.stdout.setEncoding is not a function at Object.config.out (/usr/local/lib/node_modules/requirejs/bin/r.js:31077:36) in index.html
`

Thanks for a nice plugin btw.

Enselic commented:
Build fails because of a unit test that fails, I'll fix that and add a test for my code if you think my change is reasonable.

tkareine commented:
The current plugin implementation expects the minifier to output errors to STDERR, which is visible in the terminal. But I understand that some minifiers might output errors to STDOUT. It makes sense to include a part of minifier's STDOUT output in the plugin's error message.

So, I think this is a good idea. But I wouldn't include the whole minifier STDOUT output, because it might be possible for a minifier to produce valid output before outputting its error message. How about including the last max 100 lines or so of the output? That should include enough useful information for debugging.

I plan to update the project's tests for Jekyll 3 when I get the time.

Enselic commented:
Sounds reasonable. I have now truncated the output to include just the 5000 last chars of output. Test cases also modified and added.

I went with 5000 chars instead of 100 lines because of ease of implementation. 5000 chars should be enough for a big stack trace for example.

Enselic commented:
Build passes locally for me but fails on ci server. Yay. I'll investigate when I get time. I probably have to spam new commits with different approaches and see what works...

Anyways, eager to hear your comments on my last commit.

tkareine commented:
Thanks! I clarified logging (Jekyll and Liquid remove newlines from exceptions) and made file reading to be more memory efficient. I'll be busy with other things the rest of the week, but when I'll get the time, I'll test these changes more and merge them to master if everything's ok.

tkareine commented:
I merged the changes are into the master branch, thanks!

Enselic commented:
No problem, my pleasure. Thanks for improving the code
